cloud/aws/actuators/machine/actuator.go

```diff
@@ -1,63 +1,504 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
 
 package machine
 
 import (
+	"encoding/base64"
 	"fmt"
 
 	"github.com/golang/glog"
+	log "github.com/sirupsen/logrus"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	providerconfigv1 "sigs.k8s.io/cluster-api-provider-aws/cloud/aws/providerconfig/v1alpha1"
 	clusterv1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+	clusterclient "sigs.k8s.io/cluster-api/pkg/client/clientset_generated/clientset"
 	client "sigs.k8s.io/cluster-api/pkg/client/clientset_generated/clientset/typed/cluster/v1alpha1"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+
+	clustoplog "sigs.k8s.io/cluster-api-provider-aws/cloud/aws/logging"
+)
+
+const (
+	awsCredsSecretIDKey       = "awsAccessKeyId"
+	awsCredsSecretAccessKey   = "awsSecretAccessKey"
+	userDataSecretKey         = "userDataKey"
+	ec2InstanceIDNotFoundCode = "InvalidInstanceID.NotFound"
 )
 
-// Actuator is responsible for performing machine reconciliation
+// Actuator is the AWS-specific actuator for the Cluster API machine controller
 type Actuator struct {
-	clusterClient client.ClusterInterface
+	kubeClient    kubernetes.Interface
+	clusterClient clusterclient.Interface
+	logger        *log.Entry
+	clientBuilder func(kubeClient kubernetes.Interface, secretName, namespace, region string) (Client, error)
 }
 
 // ActuatorParams holds parameter information for Actuator
 type ActuatorParams struct {
 	ClusterClient client.ClusterInterface
 }
 
-// NewActuator creates a new Actuator
-func NewActuator(params ActuatorParams) (*Actuator, error) {
-	return &Actuator{
-		clusterClient: params.ClusterClient,
-	}, nil
+// NewActuator returns a new AWS Actuator
+func NewActuator(kubeClient kubernetes.Interface, clusterClient clusterclient.Interface, logger *log.Entry) (*Actuator, error) {
+	actuator := &Actuator{
+		kubeClient:    kubeClient,
+		clusterClient: clusterClient,
+		logger:        logger,
+		clientBuilder: NewClient,
+	}
+	return actuator, nil
 }
 
-// Create creates a machine and is invoked by the Machine Controller
+// Create runs a new EC2 instance
 func (a *Actuator) Create(cluster *clusterv1.Cluster, machine *clusterv1.Machine) error {
-	glog.Infof("Creating machine %v for cluster %v.", machine.Name, cluster.Name)
-	return fmt.Errorf("TODO: Not yet implemented")
+	mLog := clustoplog.WithMachine(a.logger, machine)
+	mLog.Info("creating machine")
+	instance, err := a.CreateMachine(cluster, machine)
+	if err != nil {
+		mLog.Errorf("error creating machine: %v", err)
+		return err
+	}
+
+	return a.updateStatus(machine, instance, mLog)
```

Comment:
Part of the status that gets updated when the machine gets created is the PublicIP. However, after a call to runInstance, most of the time a PublicIP is not yet allocated. If we don't yet have complete status (ie. the instance state is Pending, instead of Running), maybe we should return an error so the machine controller keeps retrying until we have complete status we can set.

Reply:
Let's track that in the follow-up PR. This will need slight re-desigin as you described in the comment above.